File: C/my_c/reader.c
#include"reader.h"
#include"types.h"

#include<stdlib.h>
#include<stdio.h>
#include<string.h>
#include <glib-2.0/glib.h>
#include <glib-2.0/glib-object.h>
#include <glib-2.0/glib/gregex.h>
#include <glib-2.0/glib/garray.h>

//Declare
MalVal *read_form(Reader *reader);

Reader *reader_new()
{
  Reader *reader = (Reader*)MAL_GC_MALLOC(sizeof(Reader));
  reader->array = g_array_sized_new(TRUE,FALSE,sizeof(char*),8);
  reader->position = 0;
  return reader;
}

char *reader_peek(Reader *reader)
{
  return g_array_index(reader->array,char * ,reader->position);
}

Reader *tokenizer(char *line)
{

  GRegex *regex;
  GMatchInfo *matchInfo;
  GError * err = NULL;

  Reader *reader = reader_new();
  regex = g_regex_new("[\\s ,]*(~@|[\\[\\]{}()'`~@]|\"(?:[\\\\].|[^\\\\\"])*\"|;.*|[^\\s \\[\\]{}()'\"`~@,;]*)", 0, 0, &err);
  g_regex_match(regex,line,0,&matchInfo);
  if(err != NULL)
    {
      fprintf(stderr,"Tokenize error:%s\n",err->message);
      return NULL;
    }
  while(g_match_info_matches(matchInfo))
    {
      gchar *result = g_match_info_fetch(matchInfo,1);
      if(result[0] != '\0' && result[0] != ';')
        {
          reader_append(reader,result);
        }
      g_match_info_next(matchInfo,&err);
    }
  g_match_info_free(matchInfo);
  g_regex_unref(regex);
  if(reader->array->len == 0)
    {
      reader_free(reader);
      return NULL;
    }
  else
    {
      return reader;
    }

}

/* a left paren then read_list
 * otherwise read_aton is called
 * with the Reader object
 */
MalVal *read_form(Reader *reader)
{
  char *token;
  MalVal *form = NULL,*tmp;

  token = reader_peek(reader);
  if(!token){return NULL;}
  switch(token[0])
    {
    case ';':
      break;
    case '\'':
      reader_next(reader);
      form = _listX(2,malval_new_symbol("quote"),read_form(reader));
      break;
    case '`':
      reader_next(reader);
      form = 
    }
}



MalVal *read_atom(Reader *reader)
{
  char *token;
  GRegex *regex;
  GMatchInfo * matchInfo;
  GError *err = NULL;
  gint pos;
  MalVal *atom;

  token = reader_next(reader);
  regex = g_regex_new("(^-?[0-9]+$)|(^-?[0-9][0-9.]*$)|(^nil$)|(^true$)|(^false$)|^\"(.*)\"$|:(.*)|(^[^\"]*$)", 0, 0, &err);
  g_regex_match(regex,token,0,&matchInfo);

  if(g_match_info_fetch_pos(matchInfo,1,&ops,NULL) && pos != -1)
    {
      //integer
      atom = malval_new_integer(g_ascii_strtoll(token,NULL,10));
    }
  else
    {
      malval_free(atom);
      atom = NULL;
    }
  return atom;

}

/* repeatedly call read_form until a ')'
 *
 */
MalVal *read_list(Reader *reader,MalType type,char start,char end)
{
  MalVal *ast,*form;
  char *token = reader_next(reader);
  if(token[0] != start)
    {
      abort("expected '(' or '['");
    }
  ast = malval_new_list(type,g_array_new(TRUE,TRUE,sizeof(MalVal*)));
  while((token = reader_peek(reader)) && token[0] != end)
    {
      form = read_form(reader);
      if(!form)
        {
          if(!mal_error)
            {
              abort("unknown read_list failure");
            }
          g_array_free(ast);
          return NULL;
        }
      g_array_append_val(ast->val.array,form);
    }
  if(!token){ abort ("expected ')' or ']', got EOF"); }
  reader_next(reader);
  return ast;
}

/*call tokenizer create a new Reader object
 *call read_form with the Reader instance
 *
 */
MalVal *read_str(char *str)
{
  Reader *reader;
  char *token;
  MalVal *ast = NULL;
  reader = tokenizer(str);
  if(reader)
    {
      ast = read_form(reader);
      reader_free(reader);
    }
  return ast;
}

File: nodejs/mal/reader.js
var reader = {};
var types = require('./types');

function Reader(tokens){
	this.tokens = tokens.map(a => {return a;});
	this.position = 0;
}

Reader.prototype.next = function(){ return this.tokens[this.position++];};
Reader.prototype.peek = function(){ return this.tokens[this.position];};

function tokenize(str) {
    var re = /[\s,]*(~@|[\[\]{}()'`~^@]|"(?:\\.|[^\\"])*"|;.*|[^\s\[\]{}('"`,;)]*)/g;
	var results = [];
	while((match = re.exec(str)[1]) != '') {
		if( match[0] === ';') {continue;}
		results.push(match);
	}
	return results;
}

function read_str(str) {
	var tokens = tokenize(str);
	if(tokens.length === 0) {
		throw new BlankException();
	}
	return read_form(new Reader(tokens));
}

function read_form(reader) {
	var token = reader.peek();
	switch(token) {
	case ';': return null; // ignore comments
	case '\'': render.next();
		return [types._symbol('quote'), read_form(reader)];
	case '`': reader.next();
		return [types._symbol('quasiquote'), read_form(reader)];

		//list
	case ')': throw new Error("unexpected ')'");
	case '(': return read_list(reader);

		//atom
	default: return read_atom(reader);
	}
}

function BlankException(msg) {
}

function read_list(reader, start, end){
	start  =  start || '(';
	end = end || ')';
	var ast = [];
	var token = reader.next();
	if(token !== start){
		throw new Error("expected '" + start + "'");
	}
	while((token = reader.peek()) != end){
		if(!token){
			throw new Error("expected '" + end + "', got EOF");
		}
		ast.push(read_form(reader));
	}
	reader.next();
	return ast;
}

function read_atom(reader) {
	var token = reader.next();
	if(token.match(/^-?[0-9]+$/)) {
		return parseInt(token,10);
	} else if(token.match(/^-?[0-9][0-9.]*$/)) {
		return parseFloat(token, 10);
	} else if(token[0] === "\"") {
		return token.slice(1, token.length-1)
			.replace(/\\"/g,'"')
			.replace(/\\n/g,"\n")
			.replace(/\\\\/g,"\\");
	} else if(token[0] === ":") {
		return types._keyword(token.slice(1));
	} else if(token === "nil") {
		return null;
	} else if(token === "true") {
		return true;
	} else if(token === "false") {
		return false;
	} else {
		return types._symbol(token);
	}
}

exports.Reader = reader.Reader = Reader;
exports.BlankException = reader.BlankException = BlankException;
exports.tokenize = reader.tokenize = tokenize;
exports.read_form = reader.read_form = read_form;
exports.read_str = reader.read_str = read_str;
